mountpoint-s3/src/fs.rs

```diff
@@ -726,10 +698,29 @@ where
             return Err(err!(libc::EINVAL, "O_SYNC and O_DSYNC are not supported"));
         }
 
-        // All file handles will be lazy initialized on first read/write.
-        let state = FileHandleState::new(lookup, flags, pid).await.into();
+        let state = if flags & libc::O_RDWR != 0 {
+            let is_truncate = flags & libc::O_TRUNC != 0;
+            if !remote_file || (self.config.allow_overwrite && is_truncate) {
```

Comment:
The current behavior on open of an existing file is (with allow_overwrite):
* O_WRONGLY: succeed with O_TRUNC, fail otherwise,
* O_RDWR: open as write with O_TRUNC, as read without.

Before this change, the intended behavior was same, but enforced on first read/write (or flush or release!). This allowed for the edge case of open without O_TRUNC and no writes (or reads), which would result in a no op. After this change, this is resolved on open and it's either an error or a read.

Reply:
Looks like this test was failing before and now passes:

```rust


fn overwrite_empty_file<F>(creator_fn: F, prefix: &str)
where
    F: FnOnce(&str, TestSessionConfig) -> (TempDir, BackgroundSession, TestClientBox),
{
    let filesystem_config = S3FilesystemConfig {
        allow_overwrite: true,
        ..Default::default()
    };
    let test_config = TestSessionConfig {
        filesystem_config,
        ..Default::default()
    };
    let (mount_point, _session, mut test_client) = creator_fn(prefix, test_config);

    // Make sure there's an existing directory and a file
    test_client.put_object("dir/hello.txt", b"").unwrap();

    let _subdir = mount_point.path().join("dir");
    let path = mount_point.path().join("dir/hello.txt");

    // Open the file in non-truncate mode and do nothing
    let mut options = File::options();
    let _err = options.write(true).open(&path).expect_err("open should not succeed without O_TRUNC");
}

#[cfg(feature = "s3_tests")]
#[test]
fn overwrite_empty_file_s3() {
    overwrite_empty_file(fuse::s3_session::new, "overwrite_empty_file");
}

#[test]
fn overwrite_empty_file_mock() {
    overwrite_empty_file(fuse::mock_session::new, "overwrite_empty_file");
}
```

Reply:
> the intended behavior was same
sounds good and thanks for explanation why making initialisation eager, I definitely lost the original intention of fixing read-stale-after-no-op-truncate

Reply:
I've pushed a test similar to above which verifies the new behavior. I think we should adopt the new behavior, and shouldn't have allowed opening a write handle on an empty file.

This change brings the behavior in line both with and without `--allow-overwrites`.

Reply:
> Looks like this test was failing before and now passes:

Wow, you've quickly assembled a test case for that! So you're opening  in`O_WRONLY`-non-truncate mode and it's failing fast. A little different from my original thought of `O_RDWR`-non-truncate mode. In both cases if the error was previously reported on `write` and now it's just reported earlier, we're good in my opinion.